File: tests/typesafety/typescript/payment_states.ts
// expect: passes

type Pending = { readonly kind: "pending"; readonly requestedAt: string };
type Settled = { readonly kind: "settled"; readonly requestedAt: string; readonly receiptId: string };
type Payment = Pending | Settled;

function assertNever(value: never): never {
  throw new Error(`unhandled case: ${JSON.stringify(value)}`);
}

export function describe(payment: Payment): string {
  switch (payment.kind) {
    case "pending":
      return `waiting since ${payment.requestedAt}`;
    case "settled":
      return `settled, receipt ${payment.receiptId}`;
    default:
      return assertNever(payment);
  }
}

File: tests/typesafety/python/payment_states.py
# expect: passes
# run: yes
# title: Pending and Settled become separate types
# improves: payment_optional
from dataclasses import dataclass
from typing import assert_never


@dataclass(frozen=True)
class Pending:
    requested_at: str


@dataclass(frozen=True)
class Settled:
    requested_at: str
    receipt_id: str


type Payment = Pending | Settled


def describe(payment: Payment) -> str:
    match payment:
        case Pending(requested_at=at):
            return f"waiting since {at}"
        case Settled(receipt_id=receipt):
            return f"settled, receipt {receipt}"
        case _:
            assert_never(payment)


assert describe(Pending("09:00")) == "waiting since 09:00"
assert describe(Settled("09:00", "r-42")) == "settled, receipt r-42"
